proposals/tech/pt/optimize_stored_procedure/README.md

```diff
@@ -0,0 +1,191 @@
+( 
+    @BusinessEntityID  INT = NULL,
+    @StateProvinceID   INT = NULL,
+    @CountryRegionCode NVARCHAR(3) = NULL
+)
+AS
+BEGIN
+
+	SET NOCOUNT ON 
+
+	SELECT
+		A.FirstName, 
+		A.LastName
+	 FROM Person.Person as A 
+	 LEFT OUTER JOIN Person.BusinessEntityAddress as B 
+       ON A.BusinessEntityID = B.BusinessEntityID 
+	 LEFT OUTER JOIN Person.Address as C 
+	   ON B.AddressID = C.AddressID 
+	 LEFT OUTER JOIN Person.StateProvince as D 
+	   ON C.StateProvinceID = D.StateProvinceID 
+	 LEFT OUTER JOIN Person.CountryRegion as E 
+	   ON D.CountryRegionCode = E.CountryRegionCode 
+	WHERE (@BusinessEntityID IS NULL OR A.BusinessEntityID = @BusinessEntityID) 
+	  AND (@StateProvinceID IS NULL OR C.StateProvinceID = @StateProvinceID) 
+	  AND (@CountryRegionCode IS NULL OR D.CountryRegionCode = @CountryRegionCode) 
+	OPTION (RECOMPILE)
+
+END 
+```
+
+
+Ao chamar novamente a procedure um plano de execução mais performático foi escolhido:
+
+![Plano ótimo selecionado](./storedprocedure3.png)
+
+
+## **Usando *Dynamic Queries* para otimizar suas *Stored Procedures***
+
+Apesar da escrita ser mais trabalhosa (e um pouco menos legível), utilizar *dynamic queries* (consultas dinâmicas) para implementar a *procedure* acima, pode resolver os problemas de otimização de código e *parameter sniffing* de uma forma permanente e mais eficaz.
+
+```
+CREATE   PROCEDURE [dbo].[usp_GetPersonOptimized]
+(
+	@BusinessEntityID	INT = NULL,
+	@StateProvinceID	INT = NULL,
+	@CountryRegionCode	NVARCHAR(3) = NULL
+)
+AS
+BEGIN
+
+	SET NOCOUNT ON
+
+	declare @query nvarchar(max) = '
+	SELECT
+		A.FirstName,
+		A.LastName
+	FROM Person.Person as A ' +
+	IIF(@StateProvinceID IS NULL AND @CountryRegionCode IS NULL, '', '
+	INNER JOIN Person.BusinessEntityAddress as B ON A.BusinessEntityID = B.BusinessEntityID 
+	INNER JOIN Person.Address as C ON B.AddressID = C.AddressID
+	INNER JOIN Person.StateProvince as D ON C.StateProvinceID = D.StateProvinceID ') +
+	IIF(@CountryRegionCode IS NULL, '', '
+	INNER JOIN Person.CountryRegion as E ON D.CountryRegionCode = E.CountryRegionCode ') + '
+	WHERE 1 = 1 ' +
+	IIF (@BusinessEntityID IS NULL, '', '
+	  AND A.BusinessEntityID = @BusinessEntityID ') +
+	IIF (@StateProvinceID IS NULL, '', '
+	  AND C.StateProvinceID = @StateProvinceID ') +
+	IIF (@CountryRegionCode IS NULL, '', '
+	  AND D.CountryRegionCode = @CountryRegionCode ')
+
+	EXEC sp_executesql
+		@stmt			= @query,
+		@params			= N'@BusinessEntityID INT, @StateProvinceID INT, @CountryRegionCode NVARCHAR(3)',
+		@BusinessEntityID	= @BusinessEntityID,				
+		@StateProvinceID	= @StateProvinceID,					
+		@CountryRegionCode	= @CountryRegionCode
+
+END
+GO
+```
+
+>O uso do filtro *WHERE* 1 = 1, é apenas um facilitador para a escrita dos demais filtros da consulta. Dessa forma não se faz necessário ficar testando se algum filtro já foi adicionado anteriormente a consulta. Na hora de avaliar os predicados da consulta o compilador ignora esse filtro ao montar o plano de execução.
```

Comment:
Por que facilita?

Reply:
caso eu não use ele, teria que ficar testando se eu já concatenei a cláusula WHERE antes de escrever os outros ANDs.
algo assim:
'
INNER JOIN Person.CountryRegion as E ON D.CountryRegionCode = E.CountryRegionCode ') +
IIF (@BusinessEntityID IS NULL, '', '
WHERE A.BusinessEntityID = @BusinessEntityID ') +
IIF (@StateProvinceID IS NULL, '', IIF(@BusinessEntityID IS NULL, 'WHERE ', 'AND ') + '
	  C.StateProvinceID = @StateProvinceID ') +
IIF (@CountryRegionCode IS NULL, '', IIF(@BusinessEntityID IS NULL OR @StateProvinceID IS NULL, 'WHERE ', 'AND ') + '
	  D.CountryRegionCode = @CountryRegionCode ')

O que de certa forma deixa menos legível o código.
Então eu opto por usar um where sem sentido (1 = 1), para uma escrita mais clean.